### ISSUE TITLE
Reduce Apveyor build time by over an hour by hosting miktex elsewhere

### ISSUE THREAD

**Issue by cmorty** [doxygen/doxygen#7010]
Downloading miktex from http://doxygen.nl/testing/miktex.zip takes about 10 min (See https://ci.appveyor.com/project/doxygen/doxygen/builds/24779389/job/w5evau97788f5b8g#L31 and hold the mouse cursor over the lines). Hosting this elsewhere , e.g. at github, probably using a sperate repo, can speed up the total test by over an hour.

**Comment by cmorty** [comment]
As an alternative the build cache could be used: https://www.appveyor.com/docs/build-cache/

**Comment by lrineau** [comment]
[**@cmorty**](https://github.com/cmorty) commented on [May 24, 2019, 9:34 AM GMT+2](https://github.com/doxygen/doxygen/issues/7010#issue-448021877 "2019-05-24T07:34:52Z - Replied by Github Reply Comments"):
> Downloading miktex from [http://doxygen.nl/testing/miktex.zip](http://doxygen.nl/testing/miktex.zip) takes about 10 min (See [ci.appveyor.com/project/doxygen/doxygen/builds/24779389/job/w5evau97788f5b8g#L31](https://ci.appveyor.com/project/doxygen/doxygen/builds/24779389/job/w5evau97788f5b8g#L31) and hold the mouse cursor over the lines). Hosting this elsewhere , e.g. at github, probably using a sperate repo, can speed up the total test by over an hour.

You read the numbers wrong. The download, line 31, took 9 second, and that is the unzipping using 7z that took 10 minutes, line 32!

**Comment by lrineau** [comment]
You should get MikTEX [from Chocolatey][1], and [use the cache as instructed by AppVeyor doc][2].

[1]: https://chocolatey.org/packages?q=MiKTeX
[2]: https://www.appveyor.com/docs/build-cache/#caching-chocolatey-packages

**Comment by cmorty** [comment]
@lrineau Nope, I didn't read them wrong. The Times show when the instruction was executed (Which is kind of is what all logging systems do). I even watched it more or less live.

**Comment by doxygen** [comment]
Storing miktex.zip on doxygen.nl was a quick hack to keep things running. I agree a better solution is needed. Any help getter this data cached would be welcome. I'm not that familiar with AppVeyor or/and Chocolatey.

**Comment by cmorty** [comment]
I'm currently looking into this. The problem is, that ctan only mirrors the latest version of miktex and the filename changes. Thus the link needs to be updated with every update. And chocolaty is not an option either as the packages are not up to date.

**Comment by albert-github** [comment]
@cmorty another problem with LaTeX (and thus MikTex) at the moment is the problem in the versions after mid of December 2018 in respect to the tabu package (see also: https://github.com/tabu-fixed/tabu), so using a cached fixed version (as the version on doxygen.nl is) is currently a good choice.
Having the package available in a faster way definitely would be a big advantage.

**Comment by cmorty** [comment]
Ok, since yesterday there is a static link: https://github.com/MiKTeX/miktex/issues/321

I've tested it, and it seems to work. Installation still takes over a minute, but it's still better than the 10 min download: https://ci.appveyor.com/project/cmorty/doxygen/build/job/tx6sx1jo46mpjfx8

I'll see whether I can improve this further.

**Comment by albert-github** [comment]
I see that the first test succeeded and that all test cases succeeded.
Time used goes for the first test from 22 min -> 16 min, so a nice improvement but still quite slow.

Would it also be, easy, possible to test whether the docs build (make docs), we didn't yet include them for the appveyor build because no tests were done in respect to the mentioned, possible, tabu problem.

**Comment by cmorty** [comment]
@albert-github I don't get the last sentence. Yet I'm not sure how much the whole thing can be sped up.

Setup is: about 2:30 min in total (I'm rounding as I see fit): 
* Miktex: ~ 1:20
* pip: ~30 (15s download and 15 installation)
* conan : 10 s
The rest is Building and Testing.
* cmake: 1:35 
* Build: 7:00
* Test: 5:30

Thus I see the following low hanging fruits: 
 * Miktex:  I'll give it a try using the setup tool and only installing "essential". Though this might cost a bit of additional time when packages need to be installed individually later. But caching might be an option here. 
* cmake: Adding a result cache might help a lot. I'm not familiar with cmake, pointers are wellcome
* msbuild: clcache might improve things here. I'll give it a quick shot.

Other ideas are wellcome.

**Comment by cmorty** [comment]
Oh, and strip testing for VS 2015 speeds up things by 100%, too.

**Comment by cmorty** [comment]
And something for those familiar with cmake: It seems to me that doxygen is rebuilt for the tests. Is that correct? 
https://ci.appveyor.com/project/cmorty/doxygen/builds/24884828/job/lauu9kep1ax54ga9

**Comment by albert-github** [comment]
Regarding my "last line": 

> Would it also be, easy, possible to test whether the docs build (make docs), we didn't yet include them for the appveyor build because no tests were done in respect to the mentioned, possible, tabu problem.

Currently the documentation for doxygen is not build as there were problems with the reliability of MikTeX (see comment in `appveyor.yml` file: Disabled building document due to unreliable MikTeX installation) and as the appveyor build was very slow it was also seen as a waste of resources (doxcumentation was build on Linux and this should be enough for a test).

At the moment theer are also some problems as indicated with the tabu package so it is not guaranteed that the doxygen documentation can be build with the proposed version.

This said my idea for the test was:
- would it be possible to build the documentation in this way
- what would be the time required for it

**Comment by albert-github** [comment]
Regarding the double generation of doxygen in case of tests.

I see it in the appveyor log file.

The appveyor builds use the "vcxproj" approach, I don't know if this is a peculiarity here as I don't see it when I use the 'nmake' approach on Windows or use the Makefile approach on Cygwin.

When doing a build with nmake and subsequently a nmake test.
I see for the later just:
```
>nmake tests

Microsoft (R) Program Maintenance Utility Version 14.16.27024.1
Copyright (C) Microsoft Corporation.  All rights reserved.

[  1%] Built target md5
[ 21%] Built target qtools
[ 21%] Built target generate_configvalues_header
[ 26%] Built target vhdlparser
[ 34%] Built target doxycfg
[100%] Built target _doxygen
[100%] Built target doxygen
Scanning dependencies of target tests
[100%] Running doxygen tests...
1..85
ok 1 - [001_a.dox]: test the \a and \mainpage commands
...
```

so no double building at all.

**Comment by cmorty** [comment]
Context: https://github.com/doxygen/doxygen/issues/7010#issuecomment-496881107

@albert-github Ok. I think I got it: So the question is whether it's reasonable to activate the latex-based building of the documentation on Windows, right?

I can't answer that. I tried tuning on those steps and the code hung there without progress. It's likely related to a GUI-Element popping up. I'll look into it.

**Comment by albert-github** [comment]
@cmorty 

Indeed this was the question (it is not only the LaTeX documentation that is build but also the HTML documentation, there is no separate target for the HTML documentation though).

Indeed the problem might be that a required package is not installed and that MikTex tries to install these packages. There are quite a few packages, when I remember correctly, that need to be installed in this way so this might take a while and thus it is not sure if it is feasible to install them for each build but when a cached vesrion is availabe it might be worthwhile.

**Comment by albert-github** [comment]
Code has been integrated in master on github (please don't close the issue as this will be done at the moment of a release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.8.16.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
